frontend/src/store/playerElement.ts

```diff
@@ -85,6 +97,41 @@ class PlayerElementStore {
     }
   };
 
+  public toggleVisualiser = (): void => {
+    const visualiserContainer =
+      document.querySelectorAll<HTMLElement>('.visualiser')[0];
+    const audioElement = document.querySelector<HTMLMediaElement>('audio');
```

Comment:
This is a really bad practice at Vue. Besides you don't need this at all.

First of all, the visualizer itself warrants it's own component (unlike the button), where the only element is a canvas. Then, you reference that element by accessing it's `ref`. More info: https://vuejs.org/guide/essentials/template-refs.html

When that button gets clicked in the music player, you can toggle a boolean ref that toggles this component with `v-if`. in the `onMounted` hook, you attach the `AudioMotionAnalyzer` DOM, while on `onBeforeUnmount` or `onUnmounted` you destroy it (you don't right now, which will eventually lead to memory leaks).

For accessing the dom element of the media player, you only need to import it [from here](https://github.com/jellyfin/jellyfin-vue/blob/master/frontend/src/store/globals.ts#L20). Do note that won't work in development, just in production (as noted [here](https://github.com/jellyfin/jellyfin-vue/blob/master/frontend/src/components/Playback/PlayerElement.vue#L40)), so you must build the client using npm run build.

With that architecture, you don't need to touch the playerlement store for anything and you'd be good with this with perhaps 50 lines only.

Reply:
Thanks for the help, I'm aware my code is pretty awful and I really appreciate the guidance, I should be able to get this fixed now I know what to do and how to do it.